File: VB/4PosBackOffice.NET/myDataGridView.vb
Friend Class myDataGridView
    Inherits DataGridView
    Friend WithEvents DataGridView1 As System.Windows.Forms.DataGridView

    Public Property CtlText As String
        Get
            Return Me.BoundText
        End Get
        Set(value As String)
            Me.BoundText = value
        End Set
    End Property

    Public Property Col As Integer
        Get
            Return Me.ColumnCount
        End Get
        Set(value As Integer)
            Me.ColumnCount = value
        End Set
    End Property

    Public Property boundColumn As String
        Set(value As String)
            Me.Columns(0).DataPropertyName = value
        End Set
        Get
            Return Me.Columns(0).DataPropertyName
        End Get
    End Property

    Public Property listField As String
        Set(value As String)
            Me.CurrentRow.DataGridView.DataMember = value
        End Set
        Get
            Return Me.CurrentRow.DataGridView.DataMember
        End Get
    End Property

    Public Property row As Integer
        Set(value As Integer)
            Me.RowCount = value
        End Set
        Get
            Return Me.RowCount
        End Get
    End Property
    Public Property BoundText As String
        Set(value As String)
            Me.DataMember = value
        End Set
        Get
            Return Me.DataMember
        End Get

    End Property

    Public Property DataField As Object
        Set(value As Object)
            Me.DataSource = value
        End Set
        Get
            Return Me.DataSource
        End Get
    End Property

    Public Sub set_RowData(ByRef rowSet As Integer, ByRef value As Object)
        Me.Rows(rowSet).Cells.Add(value)
    End Sub

    Public Property CellBackColor As Color
        Get
            Return Me.BackgroundColor
        End Get
        Set(value As Color)
            Me.BackgroundColor = value
        End Set
    End Property

    Public Sub set_TextMatrix(ByRef myRow As Integer, ByRef myColumn As Integer, ByRef myValue As String)
        Me.Rows(myRow).Cells(myColumn).Value = myValue
    End Sub

    Public Function get_TextMatrix(ByRef myRow As Integer, ByRef myCol As Integer) As String
        Return Me.Rows(myRow).Cells(myCol).Value.ToString
    End Function

    Public Sub set_RowHeight(ByRef myRow As Integer, ByRef height As Integer)
        Me.Rows(myRow).Height = height
    End Sub

    Public WriteOnly Property CellFontBold As Boolean
        Set(value As Boolean)
            If value = True Then
                Me.CurrentCell.Style.Font = New Font(Me.Font, FontStyle.Bold)
            Else
                Me.CurrentCell.Style.Font = New Font(Me.Font, FontStyle.Regular)
            End If
        End Set
    End Property

    Public Sub set_ColWidth(ByRef myCol As Integer, ByRef width As Integer)
        Me.Columns(myCol).Width = width
    End Sub

    Public Function get_ColWidth(ByRef myCol As Integer) As Integer
        Return Me.Columns(myCol).Width
    End Function

    Public Sub CtlRefresh()
        Me.Invalidate()
    End Sub

    Public WriteOnly Property CellAlignment As Integer
        Set(value As Integer)
            Select Case value
                Case 0
                    Me.CurrentCell.Style.Alignment = DataGridViewContentAlignment.TopLeft
                Case 1
                    Me.CurrentCell.Style.Alignment = DataGridViewContentAlignment.MiddleLeft
                Case 2
                    Me.CurrentCell.Style.Alignment = DataGridViewContentAlignment.BottomLeft
                Case 3
                    Me.CurrentCell.Style.Alignment = DataGridViewContentAlignment.TopCenter
                Case 4
                    Me.CurrentCell.Style.Alignment = DataGridViewContentAlignment.MiddleCenter
                Case 5
                    Me.CurrentCell.Style.Alignment = DataGridViewContentAlignment.BottomCenter
                Case 6
                    Me.CurrentCell.Style.Alignment = DataGridViewContentAlignment.TopRight
                Case 7
                    Me.CurrentCell.Style.Alignment = DataGridViewContentAlignment.MiddleRight
                Case 8
                    Me.CurrentCell.Style.Alignment = DataGridViewContentAlignment.BottomRight
                Case 9
                    Me.CurrentCell.Style.Alignment = DataGridViewContentAlignment.NotSet
            End Select
        End Set
    End Property

    Public WriteOnly Property Alignment As Integer
        Set(value As Integer)
            Select Case value
                Case 0
                    Me.Columns(0).DefaultCellStyle.Alignment = DataGridViewContentAlignment.TopLeft
                Case 1
                    Me.Columns(0).DefaultCellStyle.Alignment = DataGridViewContentAlignment.MiddleLeft
                Case 2
                    Me.Columns(0).DefaultCellStyle.Alignment = DataGridViewContentAlignment.BottomLeft
                Case 3
                    Me.Columns(0).DefaultCellStyle.Alignment = DataGridViewContentAlignment.TopCenter
                Case 4
                    Me.Columns(0).DefaultCellStyle.Alignment = DataGridViewContentAlignment.MiddleCenter
                Case 5
                    Me.Columns(0).DefaultCellStyle.Alignment = DataGridViewContentAlignment.BottomCenter
                Case 6
                    Me.Columns(0).DefaultCellStyle.Alignment = DataGridViewContentAlignment.TopRight
                Case 7
                    Me.Columns(0).DefaultCellStyle.Alignment = DataGridViewContentAlignment.MiddleRight
                Case 8
                    Me.Columns(0).DefaultCellStyle.Alignment = DataGridViewContentAlignment.BottomRight
                Case 9
                    Me.Columns(0).DefaultCellStyle.Alignment = DataGridViewContentAlignment.NotSet
            End Select
        End Set
    End Property

    Public Sub set_colAlignment(ByRef myCol As Integer, ByRef align As Integer)
        Select Case align
            Case 0
                Me.Columns(myCol).DefaultCellStyle.Alignment = DataGridViewContentAlignment.TopLeft
            Case 1
                Me.Columns(myCol).DefaultCellStyle.Alignment = DataGridViewContentAlignment.MiddleLeft
            Case 2
                Me.Columns(myCol).DefaultCellStyle.Alignment = DataGridViewContentAlignment.BottomLeft
            Case 3
                Me.Columns(myCol).DefaultCellStyle.Alignment = DataGridViewContentAlignment.TopCenter
            Case 4
                Me.Columns(myCol).DefaultCellStyle.Alignment = DataGridViewContentAlignment.MiddleCenter
            Case 5
                Me.Columns(myCol).DefaultCellStyle.Alignment = DataGridViewContentAlignment.BottomCenter
            Case 6
                Me.Columns(myCol).DefaultCellStyle.Alignment = DataGridViewContentAlignment.TopRight
            Case 7
                Me.Columns(myCol).DefaultCellStyle.Alignment = DataGridViewContentAlignment.MiddleRight
            Case 8
                Me.Columns(myCol).DefaultCellStyle.Alignment = DataGridViewContentAlignment.BottomRight
            Case 9
                Me.Columns(myCol).DefaultCellStyle.Alignment = DataGridViewContentAlignment.NotSet
        End Select
    End Sub

    Public Function get_RowData(ByRef myRow) As Integer
        Return Me.Rows(myRow).Cells(0).Value
    End Function

    Public ReadOnly Property CellWidth As Integer
        Get
            Dim r As Rectangle
            r = Me.GetCellDisplayRectangle(Me.CurrentCell.ColumnIndex, Me.CurrentCell.RowIndex, False)
            Return r.Width
        End Get
    End Property

    Public ReadOnly Property CellLeft As Integer
        Get
            Dim r As Rectangle
            r = Me.GetCellDisplayRectangle(Me.CurrentCell.ColumnIndex, Me.CurrentCell.RowIndex, False)
            Return r.Left
        End Get
    End Property

    Public ReadOnly Property CellHeight As Integer
        Get
            Dim r As Rectangle
            r = Me.GetCellDisplayRectangle(Me.CurrentCell.ColumnIndex, Me.CurrentCell.RowIndex, False)
            Return r.Height
        End Get
    End Property

    Public ReadOnly Property CellTop As Integer
        Get
            Dim r As Rectangle
            Me.GetCellDisplayRectangle(Me.CurrentCell.ColumnIndex, Me.CurrentCell.RowIndex, False)
            Return r.Top
        End Get
    End Property

    Public Function get_RowHeight(ByRef rowID As Integer) As Integer
        Dim rs As Rectangle = Me.GetRowDisplayRectangle(rowID, False)
        Return rs.Height
    End Function

    Public Property FixedRow As Integer
        Get
            Return Me.CurrentRow.State
        End Get
        Set(value As Integer)
            Me.CurrentRow.Resizable = value
        End Set
    End Property

    Public Property FixedCols As Integer
        Get
            Return Me.Columns(0).State
        End Get
        Set(value As Integer)
            Me.Columns(0).Resizable = value
        End Set
    End Property

    Public Property TopRow As Integer
        Get

        End Get
        Set(value As Integer)

        End Set
    End Property

    Public Property FixedRows As Integer
        Get
            Dim rc As Integer = 0
            Dim c As Integer
            For c = 0 To Me.RowCount
                If Me.CurrentRow.State <> DataGridViewElementStates.Resizable Then
                    rc = rc + 1
                End If
            Next
            Return rc
        End Get
        Set(value As Integer)
            Me.CurrentRow.Resizable = value
        End Set
    End Property
    Public Property AllowAddNew As Boolean
        Get
            Return Me.AllowUserToAddRows
        End Get
        Set(value As Boolean)
            Me.AllowUserToAddRows = value
        End Set
    End Property
    Private Sub InitializeComponent()
        Me.DataGridView1 = New System.Windows.Forms.DataGridView()
        CType(Me.DataGridView1, System.ComponentModel.ISupportInitialize).BeginInit()
        CType(Me, System.ComponentModel.ISupportInitialize).BeginInit()
        Me.SuspendLayout()
        '
        'DataGridView1
        '
        Me.DataGridView1.ColumnHeadersHeightSizeMode = System.Windows.Forms.DataGridViewColumnHeadersHeightSizeMode.AutoSize
        Me.DataGridView1.Location = New System.Drawing.Point(0, 0)
        Me.DataGridView1.Name = "DataGridView1"
        Me.DataGridView1.Size = New System.Drawing.Size(240, 150)
        Me.DataGridView1.TabIndex = 0
        CType(Me.DataGridView1, System.ComponentModel.ISupportInitialize).EndInit()
        CType(Me, System.ComponentModel.ISupportInitialize).EndInit()
        Me.ResumeLayout(False)

    End Sub
End Class

File: C-Sharp/myDataGridView.cs
using CrystalDecisions.CrystalReports.Engine;
using CrystalDecisions.Windows.Forms;
using DpSdkEngLib;
using DPSDKOPSLib;
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Drawing;
using System.Diagnostics;
using System.Windows.Forms;
using System.Linq;
using System.Xml.Linq;
namespace _4PosBackOffice.NET
{
	internal class myDataGridView : DataGridView
	{

		internal System.Windows.Forms.DataGridView DataGridView1;
		public string CtlText {
			get { return this.BoundText; }
			set { this.BoundText = value; }
		}

		public int Col {
			get { return this.ColumnCount; }
			set { this.ColumnCount = value; }
		}

		public string boundColumn {
			get { return this.Columns[0].DataPropertyName; }
			set { this.Columns[0].DataPropertyName = value; }
		}

		public string listField {
			get { return this.CurrentRow.DataGridView.DataMember; }
			set { this.CurrentRow.DataGridView.DataMember = value; }
		}

		public int row {
			get { return this.RowCount; }
			set { this.RowCount = value; }
		}
		public string BoundText {
			get { return this.DataMember; }
			set { this.DataMember = value; }
		}


		public object DataField {
			get { return this.DataSource; }
			set { this.DataSource = value; }
		}

		public void set_RowData(ref int rowSet, ref object value)
		{
			this.Rows[rowSet].Cells.Add(value);
		}

		public Color CellBackColor {
			get { return this.BackgroundColor; }
			set { this.BackgroundColor = value; }
		}

		public void set_TextMatrix(ref int myRow, ref int myColumn, ref string myValue)
		{
			this.Rows[myRow].Cells[myColumn].Value = myValue;
		}

		public string get_TextMatrix(ref int myRow, ref int myCol)
		{
			return this.Rows[myRow].Cells[myCol].Value.ToString();
		}

		public void set_RowHeight(ref int myRow, ref int height)
		{
			this.Rows[myRow].Height = height;
		}

		public bool CellFontBold {
			set {
				if (value == true) {
					this.CurrentCell.Style.Font = new Font(this.Font, FontStyle.Bold);
				} else {
					this.CurrentCell.Style.Font = new Font(this.Font, FontStyle.Regular);
				}
			}
		}

		public void set_ColWidth(ref int myCol, ref int width)
		{
			this.Columns[myCol].Width = width;
		}

		public int get_ColWidth(ref int myCol)
		{
			return this.Columns[myCol].Width;
		}

		public void CtlRefresh()
		{
			this.Invalidate();
		}

		public int CellAlignment {
			set {
				switch (value) {
					case 0:
						this.CurrentCell.Style.Alignment = DataGridViewContentAlignment.TopLeft;
						break;
					case 1:
						this.CurrentCell.Style.Alignment = DataGridViewContentAlignment.MiddleLeft;
						break;
					case 2:
						this.CurrentCell.Style.Alignment = DataGridViewContentAlignment.BottomLeft;
						break;
					case 3:
						this.CurrentCell.Style.Alignment = DataGridViewContentAlignment.TopCenter;
						break;
					case 4:
						this.CurrentCell.Style.Alignment = DataGridViewContentAlignment.MiddleCenter;
						break;
					case 5:
						this.CurrentCell.Style.Alignment = DataGridViewContentAlignment.BottomCenter;
						break;
					case 6:
						this.CurrentCell.Style.Alignment = DataGridViewContentAlignment.TopRight;
						break;
					case 7:
						this.CurrentCell.Style.Alignment = DataGridViewContentAlignment.MiddleRight;
						break;
					case 8:
						this.CurrentCell.Style.Alignment = DataGridViewContentAlignment.BottomRight;
						break;
					case 9:
						this.CurrentCell.Style.Alignment = DataGridViewContentAlignment.NotSet;
						break;
				}
			}
		}

		public int Alignment {
			set {
				switch (value) {
					case 0:
						this.Columns[0].DefaultCellStyle.Alignment = DataGridViewContentAlignment.TopLeft;
						break;
					case 1:
						this.Columns[0].DefaultCellStyle.Alignment = DataGridViewContentAlignment.MiddleLeft;
						break;
					case 2:
						this.Columns[0].DefaultCellStyle.Alignment = DataGridViewContentAlignment.BottomLeft;
						break;
					case 3:
						this.Columns[0].DefaultCellStyle.Alignment = DataGridViewContentAlignment.TopCenter;
						break;
					case 4:
						this.Columns[0].DefaultCellStyle.Alignment = DataGridViewContentAlignment.MiddleCenter;
						break;
					case 5:
						this.Columns[0].DefaultCellStyle.Alignment = DataGridViewContentAlignment.BottomCenter;
						break;
					case 6:
						this.Columns[0].DefaultCellStyle.Alignment = DataGridViewContentAlignment.TopRight;
						break;
					case 7:
						this.Columns[0].DefaultCellStyle.Alignment = DataGridViewContentAlignment.MiddleRight;
						break;
					case 8:
						this.Columns[0].DefaultCellStyle.Alignment = DataGridViewContentAlignment.BottomRight;
						break;
					case 9:
						this.Columns[0].DefaultCellStyle.Alignment = DataGridViewContentAlignment.NotSet;
						break;
				}
			}
		}

		public void set_colAlignment(ref int myCol, ref int align)
		{
			switch (align) {
				case 0:
					this.Columns[myCol].DefaultCellStyle.Alignment = DataGridViewContentAlignment.TopLeft;
					break;
				case 1:
					this.Columns[myCol].DefaultCellStyle.Alignment = DataGridViewContentAlignment.MiddleLeft;
					break;
				case 2:
					this.Columns[myCol].DefaultCellStyle.Alignment = DataGridViewContentAlignment.BottomLeft;
					break;
				case 3:
					this.Columns[myCol].DefaultCellStyle.Alignment = DataGridViewContentAlignment.TopCenter;
					break;
				case 4:
					this.Columns[myCol].DefaultCellStyle.Alignment = DataGridViewContentAlignment.MiddleCenter;
					break;
				case 5:
					this.Columns[myCol].DefaultCellStyle.Alignment = DataGridViewContentAlignment.BottomCenter;
					break;
				case 6:
					this.Columns[myCol].DefaultCellStyle.Alignment = DataGridViewContentAlignment.TopRight;
					break;
				case 7:
					this.Columns[myCol].DefaultCellStyle.Alignment = DataGridViewContentAlignment.MiddleRight;
					break;
				case 8:
					this.Columns[myCol].DefaultCellStyle.Alignment = DataGridViewContentAlignment.BottomRight;
					break;
				case 9:
					this.Columns[myCol].DefaultCellStyle.Alignment = DataGridViewContentAlignment.NotSet;
					break;
			}
		}

		public int get_RowData(ref myRow)
		{
			return this.Rows[myRow].Cells[0].Value;
		}

		public int CellWidth {
			get {
				Rectangle r = default(Rectangle);
				r = this.GetCellDisplayRectangle(this.CurrentCell.ColumnIndex, this.CurrentCell.RowIndex, false);
				return r.Width;
			}
		}

		public int CellLeft {
			get {
				Rectangle r = default(Rectangle);
				r = this.GetCellDisplayRectangle(this.CurrentCell.ColumnIndex, this.CurrentCell.RowIndex, false);
				return r.Left;
			}
		}

		public int CellHeight {
			get {
				Rectangle r = default(Rectangle);
				r = this.GetCellDisplayRectangle(this.CurrentCell.ColumnIndex, this.CurrentCell.RowIndex, false);
				return r.Height;
			}
		}

		public int CellTop {
			get {
				Rectangle r = default(Rectangle);
				this.GetCellDisplayRectangle(this.CurrentCell.ColumnIndex, this.CurrentCell.RowIndex, false);
				return r.Top;
			}
		}

		public int get_RowHeight(ref int rowID)
		{
			Rectangle rs = this.GetRowDisplayRectangle(rowID, false);
			return rs.Height;
		}

		public int FixedRow {
			get { return this.CurrentRow.State; }
			set { this.CurrentRow.Resizable = value; }
		}

		public int FixedCols {
			get { return this.Columns[0].State; }
			set { this.Columns[0].Resizable = value; }
		}

		public int TopRow {

			get { }

			set { }
		}

		public int FixedRows {
			get {
				int rc = 0;
				int c = 0;
				for (c = 0; c <= this.RowCount; c++) {
					if (this.CurrentRow.State != DataGridViewElementStates.Resizable) {
						rc = rc + 1;
					}
				}
				return rc;
			}
			set { this.CurrentRow.Resizable = value; }
		}
		public bool AllowAddNew {
			get { return this.AllowUserToAddRows; }
			set { this.AllowUserToAddRows = value; }
		}
		private void InitializeComponent()
		{
			this.DataGridView1 = new System.Windows.Forms.DataGridView();
			((System.ComponentModel.ISupportInitialize)this.DataGridView1).BeginInit();
			((System.ComponentModel.ISupportInitialize)this).BeginInit();
			this.SuspendLayout();
			//
			//DataGridView1
			//
			this.DataGridView1.ColumnHeadersHeightSizeMode = System.Windows.Forms.DataGridViewColumnHeadersHeightSizeMode.AutoSize;
			this.DataGridView1.Location = new System.Drawing.Point(0, 0);
			this.DataGridView1.Name = "DataGridView1";
			this.DataGridView1.Size = new System.Drawing.Size(240, 150);
			this.DataGridView1.TabIndex = 0;
			((System.ComponentModel.ISupportInitialize)this.DataGridView1).EndInit();
			((System.ComponentModel.ISupportInitialize)this).EndInit();
			this.ResumeLayout(false);

		}
	}
}
